relay-server/src/actors/processor/session.rs

```diff
@@ -0,0 +1,452 @@
+    metrics_config: SessionMetricsConfig,
+    clock_drift_processor: &ClockDriftProcessor,
+    extracted_metrics: &mut Vec<Bucket>,
+) -> bool {
+    let mut changed = false;
+    let payload = item.payload();
+    let max_secs_in_future = config.max_secs_in_future();
+    let max_session_secs_in_past = config.max_session_secs_in_past();
+
+    let mut session = match SessionAggregates::parse(&payload) {
+        Ok(session) => session,
+        Err(error) => {
+            relay_log::trace!(
+                error = &error as &dyn Error,
+                "skipping invalid sessions payload"
+            );
+            return false;
+        }
+    };
+
+    if clock_drift_processor.is_drifted() {
+        relay_log::trace!("applying clock drift correction to session");
+        for aggregate in &mut session.aggregates {
+            clock_drift_processor.process_datetime(&mut aggregate.started);
+        }
+        changed = true;
+    }
+
+    // Validate timestamps
+    session.aggregates.retain(|aggregate| {
+        is_valid_session_timestamp(
+            received,
+            aggregate.started,
+            max_secs_in_future,
+            max_session_secs_in_past,
+        )
+    });
+
+    // Aftter timestamp validation, aggregates could now be empty
+    if session.aggregates.is_empty() {
+        return false;
+    }
+
+    // Validate attributes
+    match validate_attributes(&client_addr, &mut session.attributes) {
+        Err(_) => return false,
+        Ok(changed_attributes) => {
+            changed |= changed_attributes;
+        }
+    }
+
+    // Extract metrics if they haven't been extracted by a prior Relay
+    if metrics_config.is_enabled() && !item.metrics_extracted() {
+        for aggregate in &session.aggregates {
+            crate::metrics_extraction::sessions::extract_session_metrics(
+                &session.attributes,
+                aggregate,
+                client,
+                extracted_metrics,
+                metrics_config.should_extract_abnormal_mechanism(),
+            );
+            item.set_metrics_extracted(true);
+        }
+    }
+
+    // Drop the aggregate if metrics have been extracted in this or a prior Relay
+    if metrics_config.should_drop() && item.metrics_extracted() {
+        return false;
+    }
+
+    if changed {
+        let json_string = match serde_json::to_string(&session) {
+            Ok(json) => json,
+            Err(err) => {
+                relay_log::error!(error = &err as &dyn Error, "failed to serialize session");
+                return false;
+            }
+        };
+
+        item.set_payload(ContentType::Json, json_string);
+    }
+
+    true
+}
+
+/// Validates all sessions and session aggregates in the envelope, if any.
+///
+/// Both are removed from the envelope if they contain invalid JSON or if their timestamps
+/// are out of range after clock drift correction.
+pub fn process(state: &mut ProcessEnvelopeState, config: Arc<Config>) {
```

Comment:
Should we move this to the top of the file? We've been slowly changing our convention to pub-first. This doc only mentions it for `impl` blocks, but I guess the same applies to standalone functions: https://develop.sentry.dev/rust/#declaration-order

Reply:
yeah, sounds good! Will do